Low Temperature Radiant System Controls Part 1

RKStrand opened the pull request:
Pull request overview
---------------------
 - The purpose of this pull request is to incorporate some code that will allow some new control options for the radiant system.  While working on this enhancement, a significant amount of refactoring was also done to help improve the code in this model and better allow additional work that will happen in separate branches.  The branch includes code change in the main radiant system model routine, two unit tests, and documentation changes.  New test input files will be added as part of the next phase of the enhancement work (the next phase will also implement an additional feature into the code).

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Refactoring, NewFeature
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [X] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process (walkthrough upon request though Edwin has already looked through the code pretty thoroughly)
 - [X] If any diffs are expected, author must demonstrate they are justified using plots and descriptions: no diffs, changes are results neutral
 - [not a defect] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [not a defect] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [X] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label: no changes needed for this work
 - [none] If structural output changes, add to output rules file and add OutputChange label
 - [not applicable] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Create NFP for new radiant controls work

This commit includes a new document (new feature proposal).  Some of this work will eventually be done in a separate branch potentially.

RKStrand committed:
Changes based on NFP Review

Made some minor adjustments based on review comments.

RKStrand committed:
Update based on addition comments

A new feature to this work was requested and added to the NFP.  Also, more language was added regarding the need to have some documentation for the 2-D modeling capabilities for radiant systems.

RKStrand committed:
Revision and Transition to Design Doc

More changes as a result of review comments by team members.  This is now a design document.

RKStrand committed:
Merge branch 'develop' into LowTempRadControlsWorkPart1FY2020

RKStrand committed:
Merge branch 'develop' into LowTempRadControlsWorkPart1FY2020

RKStrand committed:
Merge branch 'develop' into LowTempRadControlsWorkPart1FY2020

RKStrand committed:
Addition of input options for new controls

Code changes and enhancements to allow for new controls.  Two new subroutines were added to accomplish this in a generic way.

RKStrand committed:
Merge branch 'develop' into LowTempRadControlsWorkPart1FY2020

RKStrand committed:
Refactoring of Radiant data structs

First step in refactoring some of the radiant system code.  This involves redoing the definitions in the .hh file.  A few changes were also needed in the .cc file as a result of a minor change.  The goal is to also associate some of the subroutines in this file with some of the specific structs as well to make the code here much more OO before doing a lot of work on the controls issues (some minor coding did already take place).

RKStrand committed:
Refactoring of Main Simulation Routines

This commit looks at refactoring the main simulation routines for the various low temperature radiant system types.  There is potentially other things that could be done at this point, but this is a good "intermediate" point.  More work could be done restructing and splitting apart existing routines, though whether or not that happens has not yet been decided.

RKStrand committed:
Refactoring One Step Further

More refactoring work on the low temperature radiant system model.

RKStrand committed:
Merge branch 'develop' into LowTempRadControlsWorkPart1FY2020

RKStrand committed:
More Refactoring and One Unit Test

Some "final" refactoring and the addition of one unit tests.  Another to follow.  This commit is not working on my machine.  The unit test has a "problem" according my computer but I'm not sure why...

RKStrand committed:
More refactoring and a second unit tests

More work on the low temperature radiant system model including another unit test.   The unit test isn't working on my machine though it is not clear why it won't run.

RKStrand committed:
Correction of some unit test fails

This corrects a few unit test fails but not all of them.

RKStrand committed:
Merge branch 'develop' into LowTempRadControlsWorkPart1FY2020

Myoldmopar committed:
Added new custom_check for included cc file, left problem line in place but cleaned up other stuff

Myoldmopar committed:
Have new search tool report the line number

Myoldmopar committed:
Fix unit test cc include

RKStrand committed:
Documentation Edits for New Code

Changes and additions to the IO and Eng Refs for the modifications as part of the radiant controls enhancements

Myoldmopar commented:
CI is exceedingly happy with this.  I'm doing final lookover on the code now.

Myoldmopar commented on .decent_ci-Linux.yaml:
```diff
@@ -28,6 +28,7 @@ compilers:
       - ./scripts/dev/check_for_tabs_in_idfs.py
       - ./scripts/dev/check_for_bom_in_idfs.py
       - ./scripts/dev/verify_cmake_dirs.py
+      - ./scripts/dev/find_included_cc_files.py
```
Great idea adding a test for accidental inclusion of cc files.

Myoldmopar commented on design/FY2020/NFP-LowTempRadSysSurfTempCtrl.md:
```diff
@@ -0,0 +1,99 @@
+Low Temperature Radiant System Surface Temperature Control
```
And to be clear, this first PR is just a cleanup phase.  In future PRs point to this existing design doc, which will then be in the develop branch, or make changes as needed to it in your new branch.

Myoldmopar commented on doc/input-output-reference/src/overview/group-radiative-convective-units.tex:
```diff
@@ -1595,9 +1607,15 @@ \subsubsection{Inputs}\label{inputs-7-019}
-Operative temperature for radiant system controls is the average of Mean Air Temperature and Mean Radiant Temperature. If the user does not select a control type, \textbf{MeanAirTemperature} control is assumed by EnergyPlus. See the throttling range and control temperature schedule fields below for more information.
+\emph{Operative temperature} for radiant system controls is the average of Mean Air Temperature and Mean Radiant Temperature.  The \emph{SurfaceFaceTemperature} option allows the user to control the radiant system using the inside face surface temperature of the radiant system (the inside surface temperature).  The \emph{SurfaceInteriorTemperature} option will allow the user to control the radian ...
+
+If the user does not select a control type, \textbf{MeanAirTemperature} control is assumed by EnergyPlus. See the throttling range and control temperature schedule fields below for more information on how the setpoint temperature is established for this particular radiant system.
```
I'm not going to dwell over the doc changes for now.  I'd just say in the future, if you are making a cleanup pass on the code in a first branch, don't worry about including any doc changes so that the doc changes go in at the exact same time as the functional code changes.

Myoldmopar commented on src/EnergyPlus/LowTempRadiantSystem.hh:
```diff
@@ -124,10 +128,9 @@ namespace LowTempRadiantSystem {
     // Types
 
-    struct HydronicRadiantSystemData
+    struct RadiantSystemBaseData
```
Hooray for a base class with methods!

Myoldmopar commented on src/EnergyPlus/LowTempRadiantSystem.hh:
```diff
@@ -76,11 +76,15 @@ namespace LowTempRadiantSystem {
-    extern int const OperativeControl; // Controls system using operative temperature
-    extern int const ODBControl;       // Controls system using outside air dry-bulb temperature
-    extern int const OWBControl;       // Controls system using outside air wet-bulb temperature
+    enum class LowTempRadiantControlTypes {
```
Hooray for enum class instead of int const!

Myoldmopar commented on src/EnergyPlus/LowTempRadiantSystem.hh:
```diff
@@ -433,58 +439,10 @@ namespace LowTempRadiantSystem {
 
     Real64 SumHATsurf(int const ZoneNum); // Zone number
 
-    void ReportLowTempRadiantSystem(int const RadSysNum, // Index for the low temperature radiant system under consideration within the derived types
```
I love seeing all these global hanging functions going away.  The more objectified the code (where appropriate, of course) the better.

Myoldmopar commented on tst/EnergyPlus/unit/LowTempRadiantSystem.unit.cc:
```diff
@@ -1859,9 +1856,276 @@ TEST_F(LowTempRadiantSystemTest, LowTempRadCalcRadSysHXEffectTermTest)
+    actualResult = ElecRadSys(1).setRadiantSystemControlTemperature();
+    EXPECT_NEAR(expectedResult, actualResult, acceptibleError);
+
+}
```
Nice unit test additions!

Myoldmopar commented:
@RKStrand thanks for taking the time to do a deep OO refactor of this code.  It is so much nicer now!

I wanted to talk about next steps.  I commented in the review about how this was just a cleanup pass, yet the docs were changed, and then noticed the IDD was also changed.  But this doesn't include the added functionality, right?  I didn't do a deep dive into any new code functionality, as I was just interested in it as a refactor coming first before functional changes.

Let me know either way.  If it is just a cleanup, we may just make note of how happy CI is at this point, and keep pushing commits this same branch/PR and I can help review between individual commits.  Alternatively, we may want to pull out the IDD/doc changes and add them to the part 2 branch.  Alternatively, we just merge this as-is, knowing that more code changes will be coming soon.

RKStrand committed:
Addition of a new test suite input file

New IDF that tests the new control options for the variable flow radiant system.

Myoldmopar commented:
OK, I have a better understanding of the intention and status here.  I'll give this another once over with the knowledge that this is making functional changes.

Myoldmopar commented:
For starters, CI is really happy so far.  Debug integration is running fine showing that the new example is passing under a debug build.  Test coverage went up a smidge, which is a great extra benefit.

I built and ran the new test file locally, error file is clean.  Spot checked the results, things are reasonable.

I'll look over the code now but I anticipate this being ready to merge shortly.  If anyone else wants to review, now's your chance.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -43997,6 +43997,8 @@ ZoneHVAC:LowTemperatureRadiant:VariableFlow,
         \key OutdoorDryBulbTemperature
         \key OutdoorWetBulbTemperature
+        \key SurfaceFaceTemperature
+        \key SurfaceInteriorTemperature
```
OK, so in part 1, this is fixed for just this component model.  Got i.

Myoldmopar reviewed (approved):
This works for me.  In the next PR, if you could show these fixed controls for all the different system types on that PR, that would be nice to see.

Myoldmopar commented:
Reviewed code -- it all looks really really great.  I understand the intent of this phase of the work, and your new example appears to be working well.  I think I gave everyone enough of a chance to review here, I'm going to merge.  Thanks for the extra effort on refactoring this @RKStrand in conjunction with your actual fixes.
